tests/potential/test_pair.py

```diff
@@ -97,6 +97,150 @@ def test_derivative(self):
+        self.assertEqual(s.coeff['1','1']['r-1'].value, 2)
+        self.assertEqual(s.coeff['1','1']['r-1'].const, True)
+        self.assertEqual(s.coeff['1','1']['r-2'].value, 3)
+        self.assertEqual(s.coeff['1','1']['r-2'].const, True)
+
+        self.assertEqual(s.coeff['1','1']['knot-0'].value, 9)
+        self.assertEqual(s.coeff['1','1']['knot-0'].const, False)
+        self.assertEqual(s.coeff['1','1']['knot-1'].value, 4)
+        self.assertEqual(s.coeff['1','1']['knot-1'].const, False)
+        self.assertEqual(s.coeff['1','1']['knot-2'].value, 1)
+        self.assertEqual(s.coeff['1','1']['knot-2'].const, True)
+
+        #test invalid r and u shapes
+        r_arr = [2,3]
+        with self.assertRaises(ValueError):
+            s.from_array(pair=('1','1'), r=r_arr, u=u_arr)
+
+        r_arr = [1,2,3]
+        u_arr = [1,2]
+        with self.assertRaises(ValueError):
+            s.from_array(pair=('1','1'), r=r_arr, u=u_arr)
+
+    def test_energy(self):
+        """Test _energy method"""
+        r_arr = [1,2,3]
+        u_arr = [9,4,1]
+
+        #test diff mode
+        s = pair.Spline(types=('1',), num_knots=3)
+        s.from_array(pair=('1','1'), r=r_arr, u=u_arr)
+        u_actual = np.array([6.25,2.25,1])
+        u = s._energy(pair=('1','1'), r=[1.5,2.5,3.5])
+        np.testing.assert_allclose(u, u_actual)
+
+        #test value mode
+        s = pair.Spline(types=('1',), num_knots=3, mode='value')
+        s.from_array(pair=('1','1'), r=r_arr, u=u_arr)
+        u_actual = np.array([6.25,2.25,1])
+        u = s._energy(pair=('1','1'), r=[1.5,2.5,3.5])
+        np.testing.assert_allclose(u, u_actual)
+
+    def test_force(self):
+        """Test _force method"""
+        r_arr = [1,2,3]
+        u_arr = [9,4,1]
+
+        #test diff mode
+        s = pair.Spline(types=('1',), num_knots=3)
+        s.from_array(pair=('1','1'), r=r_arr, u=u_arr)
+        f_actual = np.array([5,3,0])
+        f = s._force(pair=('1','1'), r=[1.5,2.5,3.5])
+        np.testing.assert_allclose(f, f_actual)
+
+        #test value mode
+        s = pair.Spline(types=('1',), num_knots=3, mode='value')
+        s.from_array(pair=('1','1'), r=r_arr, u=u_arr)
+        f_actual = np.array([5,3,0])
+        f = s._force(pair=('1','1'), r=[1.5,2.5,3.5])
+        np.testing.assert_allclose(f, f_actual)
+
+    def test_derivative(self):
+        """Test _derivative method"""
+        r_arr = [1,2,3]
+        u_arr = [9,4,1]
+
+        #test diff mode
+        s = pair.Spline(types=('1',), num_knots=3)
+        s.from_array(pair=('1','1'), r=r_arr, u=u_arr)
+        d_actual = np.array([1.125,0.625,0])
+        d = s._derivative(pair=('1','1'), param='knot-1', r=[1.5,2.5,3.5])
+        np.testing.assert_allclose(d, d_actual)
+
+        #test value mode
+        s = pair.Spline(types=('1',), num_knots=3, mode='value')
+        s.from_array(pair=('1','1'), r=r_arr, u=u_arr)
+        d_actual = np.array([0.75,0.75,0])
```

Comment:
As I said, I think these should be different derivatives in the different modes. If we could work out analytically how they relate to each other, that would be a strong test.